turbopack/crates/turbopack-core/src/module_graph/chunk_group_info.rs

```diff
@@ -0,0 +1,345 @@
+impl Eq for RoaringBitmapWrapper {}
+
+impl Deref for RoaringBitmapWrapper {
+    type Target = RoaringBitmap;
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+impl DerefMut for RoaringBitmapWrapper {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+impl Serialize for RoaringBitmapWrapper {
+    fn serialize<S: Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
+        use serde::ser::Error;
+        let mut bytes = vec![];
+        self.0.serialize_into(&mut bytes).map_err(Error::custom)?;
+        serializer.serialize_bytes(bytes.as_slice())
+    }
+}
+
+impl<'de> Deserialize<'de> for RoaringBitmapWrapper {
+    fn deserialize<D: Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
+        use serde::de::Error;
+        let bytes = <&[u8]>::deserialize(deserializer)?;
+        let map = RoaringBitmap::deserialize_unchecked_from(bytes).map_err(Error::custom)?;
+        Ok(RoaringBitmapWrapper(map))
+    }
+}
+
+#[turbo_tasks::value(transparent)]
+pub struct ChunkGroupInfo(HashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper>);
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash)]
+enum ChunkGroup {
+    /// e.g. a page
+    Entry(ReadRef<RcStr>),
+    /// a module that has an incoming async edge
+    Async(ReadRef<RcStr>),
+    /// a module with a incoming non-merged isolated edge
+    Isolated(ReadRef<RcStr>),
+    /// a module with a incoming merging isolated edge
+    IsolatedMerged(
+        /* parent: */ Box<ChunkGroup>,
+        // /* parent: */ ChunkGroupHash,
+        /* merge_tag: */ RcStr,
+    ),
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+struct ChunkGroupHash(u64);
+
+impl ChunkGroup {
+    fn hashed(&self) -> ChunkGroupHash {
+        let mut hasher = DefaultHasher::new();
+        self.hash(&mut hasher);
+        ChunkGroupHash(hasher.finish())
+    }
+}
+
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+struct ChunkGroupId(u32);
+
+pub async fn compute_chunk_group_info(graph: &ModuleGraph) -> Result<Vc<ChunkGroupInfo>> {
+    let mut next_chunk_group_id = 0u32;
+    let mut chunk_groups_to_id: HashMap<ChunkGroup, ChunkGroupId> = HashMap::new();
+    let mut chunk_groups_from_id: HashMap<ChunkGroupId, ChunkGroup> = HashMap::new();
+
+    let mut module_chunk_groups: HashMap<ResolvedVc<Box<dyn Module>>, RoaringBitmapWrapper> =
+        HashMap::new();
+
+    let mut module_depth: HashMap<ResolvedVc<Box<dyn Module>>, usize> = HashMap::new();
+
+    let entries = &graph.graphs.last().unwrap().await?.entries;
+
+    graph
+        .traverse_edges_from_entry(entries, |parent, node| {
+            if let Some((parent, _)) = parent {
+                let parent_depth = *module_depth.get(&parent.module).unwrap();
+                let entry = module_depth.entry(node.module);
+                match entry {
+                    Entry::Occupied(mut e) => {
+                        if (parent_depth + 1) < *e.get() {
```

Comment:
This should never happen. If it would happen you would need to revisit the subgraph here, which you can't. But it can't happen if you use breath first visiting...

Reply:
Changed into
```rust
        .traverse_edges_from_entries_bfs(entries, |parent, node| {
            if let Some((parent, _)) = parent {
                let parent_depth = *module_depth.get(&parent.module).unwrap();
                module_depth.entry(node.module).or_insert(parent_depth + 1);
            } else {
                module_depth.insert(node.module, 0);
            };